File: ApiApplication/Services/Impl/AuditLogService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using ApiApplication.Data;
using ApiApplication.Dtos.AuditLog;
using ApiApplication.Dtos.Pagination;
using ApiApplication.Entities;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace ApiApplication.Services.Impl;

public class AuditLogService : IAuditLogService
{
    private readonly ApplicationDbContext _context;
    private readonly ILogger<AuditLogService> _logger;

    public AuditLogService(ApplicationDbContext context, ILogger<AuditLogService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<PagedResponse<AuditLogDto>> GetAuditLogsAsync(AuditLogQueryDto query)
    {
        try
        {
            var auditLogsQuery = _context.AuditLogs.AsQueryable();

            // Apply filters
            if (!string.IsNullOrWhiteSpace(query.TableName))
            {
                auditLogsQuery = auditLogsQuery.Where(x => x.TableName == query.TableName);
            }

            if (!string.IsNullOrWhiteSpace(query.Action))
            {
                auditLogsQuery = auditLogsQuery.Where(x => x.Action == query.Action);
            }

            if (!string.IsNullOrWhiteSpace(query.UserId))
            {
                auditLogsQuery = auditLogsQuery.Where(x => x.UserId == query.UserId);
            }

            if (!string.IsNullOrWhiteSpace(query.EntityId))
            {
                auditLogsQuery = auditLogsQuery.Where(x => x.EntityId == query.EntityId);
            }

            if (query.DateFrom.HasValue)
            {
                auditLogsQuery = auditLogsQuery.Where(x => x.Timestamp >= query.DateFrom.Value);
            }

            if (query.DateTo.HasValue)
            {
                var endOfDay = query.DateTo.Value.AddDays(1).AddTicks(-1);
                auditLogsQuery = auditLogsQuery.Where(x => x.Timestamp <= endOfDay);
            }

            if (!string.IsNullOrWhiteSpace(query.SearchKeyword))
            {
                var keyword = query.SearchKeyword.ToLower();
                auditLogsQuery = auditLogsQuery.Where(x =>
                    x.UserName!.ToLower().Contains(keyword) ||
                    x.IpAddress!.ToLower().Contains(keyword) ||
                    x.TableName.ToLower().Contains(keyword) ||
                    x.EntityId.ToLower().Contains(keyword)
                );
            }

            // Get total count before pagination
            var totalItems = await auditLogsQuery.CountAsync();

            // Apply pagination
            var items = await auditLogsQuery
                .OrderByDescending(x => x.Timestamp)
                .Skip((query.Page - 1) * query.PageSize)
                .Take(query.PageSize)
                .ToListAsync();
            
            var mappedItems = items.Select(x => MapToDto(x)).ToList();

            var totalPages = (int)Math.Ceiling((double)totalItems / query.PageSize);

            return new PagedResponse<AuditLogDto>
            {
                Items = mappedItems,
                Page = query.Page,
                PageSize = query.PageSize,
                TotalItems = totalItems,
                TotalPages = totalPages
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting audit logs");
            throw;
        }
    }

    public async Task<AuditLogDetailDto?> GetAuditLogDetailAsync(Guid id)
    {
        try
        {
            var auditLog = await _context.AuditLogs.FirstOrDefaultAsync(x => x.Id == id);

            if (auditLog == null)
                return null;

            return MapToDetailDto(auditLog);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting audit log detail for id: {Id}", id);
            throw;
        }
    }

    public async Task<List<AuditLogDto>> GetEntityHistoryAsync(string tableName, string entityId)
    {
        try
        {
            var logs = await _context.AuditLogs
                .Where(x => x.TableName == tableName && x.EntityId == entityId)
                .OrderByDescending(x => x.Timestamp)
                .ToListAsync();

            return logs.Select(x => MapToDto(x)).ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting entity history for table: {TableName}, entityId: {EntityId}", tableName, entityId);
            throw;
        }
    }

    public async Task<PagedResponse<AuditLogDto>> GetByUserAsync(string userId, int page = 1, int pageSize = 10)
    {
        try
        {
            var query = new AuditLogQueryDto
            {
                UserId = userId,
                Page = page,
                PageSize = pageSize
            };

            return await GetAuditLogsAsync(query);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting audit logs for user: {UserId}", userId);
            throw;
        }
    }

    public async Task<PagedResponse<AuditLogDto>> GetByTableAsync(string tableName, int page = 1, int pageSize = 10)
    {
        try
        {
            var query = new AuditLogQueryDto
            {
                TableName = tableName,
                Page = page,
                PageSize = pageSize
            };

            return await GetAuditLogsAsync(query);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting audit logs for table: {TableName}", tableName);
            throw;
        }
    }

    public async Task<PagedResponse<AuditLogDto>> GetByActionAsync(string action, int page = 1, int pageSize = 10)
    {
        try
        {
            var query = new AuditLogQueryDto
            {
                Action = action,
                Page = page,
                PageSize = pageSize
            };

            return await GetAuditLogsAsync(query);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting audit logs for action: {Action}", action);
            throw;
        }
    }

    public async Task<int> DeleteOldLogsAsync(int daysToKeep = 90)
    {
        try
        {
            var cutoffDate = DateTime.UtcNow.AddDays(-daysToKeep);

            var logsToDelete = await _context.AuditLogs
                .Where(x => x.CreatedAt < cutoffDate)
                .ToListAsync();

            if (logsToDelete.Count > 0)
            {
                _context.AuditLogs.RemoveRange(logsToDelete);
                await _context.SaveChangesAsync();
                _logger.LogInformation("Deleted {Count} old audit logs (before {CutoffDate})", logsToDelete.Count, cutoffDate);
            }

            return logsToDelete.Count;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting old audit logs");
            throw;
        }
    }

    public async Task CreateAuditLogAsync(AuditLog auditLog)
    {
        try
        {
            auditLog.Id = Guid.NewGuid();
            auditLog.CreatedAt = DateTime.UtcNow;
            auditLog.Timestamp = DateTime.UtcNow;

            _context.AuditLogs.Add(auditLog);
            await _context.SaveChangesAsync();

            _logger.LogInformation(
                "Audit log created: Table={TableName}, Action={Action}, EntityId={EntityId}, UserId={UserId}",
                auditLog.TableName,
                auditLog.Action,
                auditLog.EntityId,
                auditLog.UserId
            );
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating audit log");
            throw;
        }
    }

    /// <summary>
    /// Map AuditLog entity to DTO
    /// </summary>
    private static AuditLogDto MapToDto(AuditLog auditLog)
    {
        return new AuditLogDto
        {
            Id = auditLog.Id,
            TableName = auditLog.TableName,
            Action = auditLog.Action,
            EntityId = auditLog.EntityId,
            OldValues = auditLog.OldValues,
            NewValues = auditLog.NewValues,
            ChangedColumns = auditLog.ChangedColumns,
            UserId = auditLog.UserId,
            UserName = auditLog.UserName,
            IpAddress = auditLog.IpAddress,
            UserAgent = auditLog.UserAgent,
            Timestamp = auditLog.Timestamp,
            CreatedAt = auditLog.CreatedAt,
            UpdatedAt = auditLog.UpdatedAt,
            CreatedBy = auditLog.CreatedBy,
            UpdatedBy = auditLog.UpdatedBy
        };
    }

    /// <summary>
    /// Map AuditLog entity to detail DTO with parsed JSON
    /// </summary>
    private AuditLogDetailDto MapToDetailDto(AuditLog auditLog)
    {
        object? oldValuesObject = null;
        object? newValuesObject = null;

        try
        {
            if (!string.IsNullOrWhiteSpace(auditLog.OldValues))
            {
                oldValuesObject = JsonSerializer.Deserialize<object>(auditLog.OldValues);
            }

            if (!string.IsNullOrWhiteSpace(auditLog.NewValues))
            {
                newValuesObject = JsonSerializer.Deserialize<object>(auditLog.NewValues);
            }
        }
        catch (Exception ex)
        {
            _logger?.LogWarning(ex, "Error deserializing audit log values for id: {Id}", auditLog.Id);
        }

        return new AuditLogDetailDto
        {
            Id = auditLog.Id,
            TableName = auditLog.TableName,
            Action = auditLog.Action,
            EntityId = auditLog.EntityId,
            ChangedColumns = auditLog.ChangedColumns,
            OldValuesObject = oldValuesObject,
            NewValuesObject = newValuesObject,
            UserId = auditLog.UserId,
            UserName = auditLog.UserName,
            IpAddress = auditLog.IpAddress,
            UserAgent = auditLog.UserAgent,
            Timestamp = auditLog.Timestamp
        };
    }
}

File: web-application/src/services/auditLogService.ts
import { axiosInstance } from "@/lib/axios";
import {
  AuditLogDto,
  AuditLogDetailDto,
  AuditLogQueryDto,
  AuditLogResponseDto,
} from "@/types/audit-log";

export const auditLogService = {
  /**
   * Get paginated audit logs with filters
   */
  async getAuditLogs(query: AuditLogQueryDto): Promise<AuditLogResponseDto> {
    const params = new URLSearchParams();
    
    if (query.tableName) params.append("tableName", query.tableName);
    if (query.action) params.append("action", query.action);
    if (query.userId) params.append("userId", query.userId);
    if (query.entityId) params.append("entityId", query.entityId);
    if (query.dateFrom) params.append("dateFrom", query.dateFrom);
    if (query.dateTo) params.append("dateTo", query.dateTo);
    if (query.searchKeyword) params.append("searchKeyword", query.searchKeyword);
    if (query.page) params.append("page", query.page.toString());
    if (query.pageSize) params.append("pageSize", query.pageSize.toString());

    const res = await axiosInstance.get<AuditLogResponseDto>(
      `/api/auditlogs?${params.toString()}`
    );
    return res.data;
  },

  /**
   * Get audit log detail by ID
   */
  async getAuditLogDetail(id: string): Promise<AuditLogDetailDto> {
    const res = await axiosInstance.get<AuditLogDetailDto>(`/api/auditlogs/${id}`);
    return res.data;
  },

  /**
   * Get change history for a specific entity
   */
  async getEntityHistory(tableName: string, entityId: string): Promise<{
    tableName: string;
    entityId: string;
    history: AuditLogDto[];
  }> {
    const params = new URLSearchParams();
    params.append("tableName", tableName);
    params.append("entityId", entityId);

    const res = await axiosInstance.get<{
      tableName: string;
      entityId: string;
      history: AuditLogDto[];
    }>(`/api/auditlogs/entity-history?${params.toString()}`);
    return res.data;
  },

  /**
   * Get audit logs by user
   */
  async getByUser(userId: string, page = 1, pageSize = 10): Promise<AuditLogResponseDto> {
    const params = new URLSearchParams();
    params.append("page", page.toString());
    params.append("pageSize", pageSize.toString());

    const res = await axiosInstance.get<AuditLogResponseDto>(
      `/api/auditlogs/by-user/${userId}?${params.toString()}`
    );
    return res.data;
  },

  /**
   * Get audit logs by table name
   */
  async getByTable(tableName: string, page = 1, pageSize = 10): Promise<AuditLogResponseDto> {
    const params = new URLSearchParams();
    params.append("page", page.toString());
    params.append("pageSize", pageSize.toString());

    const res = await axiosInstance.get<AuditLogResponseDto>(
      `/api/auditlogs/by-table/${tableName}?${params.toString()}`
    );
    return res.data;
  },

  /**
   * Get audit logs by action type
   */
  async getByAction(action: string, page = 1, pageSize = 10): Promise<AuditLogResponseDto> {
    const params = new URLSearchParams();
    params.append("page", page.toString());
    params.append("pageSize", pageSize.toString());

    const res = await axiosInstance.get<AuditLogResponseDto>(
      `/api/auditlogs/by-action/${action}?${params.toString()}`
    );
    return res.data;
  },

  /**
   * Delete old audit logs (Admin only)
   */
  async deleteOldLogs(daysToKeep: number): Promise<{
    message: string;
    deletedCount: number;
  }> {
    const params = new URLSearchParams();
    params.append("daysToKeep", daysToKeep.toString());

    const res = await axiosInstance.delete<{
      message: string;
      deletedCount: number;
    }>(`/api/auditlogs/cleanup?${params.toString()}`);
    return res.data;
  },

  /**
   * Export audit logs as CSV
   */
  async exportAuditLogs(query: Omit<AuditLogQueryDto, "page" | "pageSize">): Promise<Blob> {
    const params = new URLSearchParams();
    
    if (query.tableName) params.append("tableName", query.tableName);
    if (query.action) params.append("action", query.action);
    if (query.userId) params.append("userId", query.userId);
    if (query.dateFrom) params.append("dateFrom", query.dateFrom);
    if (query.dateTo) params.append("dateTo", query.dateTo);

    const res = await axiosInstance.get(`/api/auditlogs/export?${params.toString()}`, {
      responseType: "blob",
    });
    return res.data;
  },
};
